### DISCUSSION TITLE
端口问题

### DISCUSSION THREAD

**Discussion by stanle1** [tryzealot/zealot#1528] [Q&A] [answered]
使用非443端口 部署https 在内页调整时没带端口号 怎么解决：

method=POST path=/users/sign_in format=turbo_stream controller=Devise::SessionsController action=create status=302 allocations=829 duration=2.81 view=0.00 db=0.61 location=https://z.aa.bb/ time=2024-06-04 12:46:00 +0800 ip=172.26.0.1 username=admin

实际应该是 https://z.aa.bb:666/

**Comment by stanle1** [comment]
docker 部署的

**Comment by icyleaf** [comment]
两种方法：

1. 设置环境变量 `ZEALOT_DOMAIN=foo.example.com:1234"
2. 管理面板进系统设置修改网站域名为 `foo.example.com:1234` 保存后点击右上角的重启服务生效

**Reply by stanle1** [reply]
管理面板 无效 试过了 另外的我试试。

是改配置文件 还是 设置docker的环境变量？

**Reply by icyleaf** [reply]
> 是改配置文件 还是 设置docker的环境变量？

改配置文件

> 管理面板 无效

docker 的端口映射改了吗

**Reply by stanle1** [reply]
改了

**Reply by stanle1** [reply]
我是通过nginx 反代 到本地 容器端口
![image](https://github.com/tryzealot/zealot/assets/12638025/fa49cd53-3718-44a2-bab0-2a9f89628ffc)

**Reply by stanle1** [reply]
![image](https://github.com/tryzealot/zealot/assets/12638025/7b3a07bf-2d54-41e5-b19d-3903f6a6ee4e)
所有点击按钮后要跳转的页面 都缺了端口号

**Reply by icyleaf** [reply]
如果外层还有一层反代，确认是按照教程配置的吗：https://zealot.ews.im/zh-Hans/docs/next/self-hosted/reverse-proxies

**Reply by stanle1** [reply]
一级网页是通的。感觉是不是反代的问题

**Reply by icyleaf** [reply]
看样子是用户登录和登出有问题，其他页面没看到有啥

**Reply by stanle1** [reply]
因为跳转的路径缺了端口 所以表现是出问题 实际再次刷新页面就是登陆状态了。

**Reply by stanle1** [reply]
备注一下 我这边时没有 443端口的 因为部署在 家宽带nas上。 也不可能在通过反代443转到我的目标端口上

**Reply by icyleaf** [reply]
或不会是 nginx 需要配置 `proxy_set_header Host $host:$server_port;`

**Reply by stanle1** [reply]
我试试

**Reply by stanle1** [reply]
无效。。
会不会是代码的bug?

**Reply by stanle1** [reply]
我试了改为ip 能把端口带上

**Reply by icyleaf** [reply]
我也没找到哪里的问题，尴尬了...

**Reply by icyleaf** [reply]
应该还是 nginx 的问题，我使用 caddy 2 是没有任何问题的

**Reply by icyleaf** [reply, marked as answer]
[找到方法](https://github.com/rails/rails/issues/39366#issuecomment-631493330)解决了，另外 Zealot 还需要配置环境变量 `RAILS_SERVE_STATIC_FILES=true` 或者把 `/public` URL 路径在 nginx 的 assets 静态文件规则上，否则静态资源无法正常加载。

```diff
server {
-  listen 443 ssl http2;  
+  listen 8443 ssl http2;    
-  listen [::]:443 ssl http2; 
+  listen [::]:8443 ssl http2; 

...

-    proxy_set_header  X-Forwarded-Host $host;
+    proxy_set_header  X-Forwarded-Host $host:$server_port;

...
}
```

**Reply by stanle1** [reply]
![image](https://github.com/tryzealot/zealot/assets/12638025/ba4ee5e7-0365-4a4a-8539-4b60f4961258)
![image](https://github.com/tryzealot/zealot/assets/12638025/faa036bb-3af9-4833-81b4-4a647a122b00)
![image](https://github.com/tryzealot/zealot/assets/12638025/c4873981-3488-433e-9ee3-2e99270c3c38)

重新配置后 点登陆后 明显就是页面重定向时 没带上端口。。感觉到未到nginx。

**Comment by stanle1** [comment]
![image](https://github.com/tryzealot/zealot/assets/12638025/19574ad5-8205-4b01-a5bc-f423c2521b3a)
默认我就是这样设置的
